File: src/com/secretary/helloworld/ui/activities/TaskActivity.java
package com.secretary.helloworld.ui.activities;

import android.app.Activity;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.*;

import com.secretary.helloworld.R;
import com.secretary.helloworld.Task;
import com.secretary.helloworld.AppLogger;
import com.secretary.helloworld.data.dao.TaskDao;
import com.secretary.helloworld.data.dao.CompletionDao;
import com.secretary.helloworld.ui.adapters.TaskListAdapter;
import com.secretary.helloworld.ui.dialogs.TaskEditDialog;
import com.secretary.helloworld.ui.dialogs.CompletionDialog;
import com.secretary.helloworld.utils.StatsManager;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Simplified Task Activity using extracted components.
 * Reduced from 971 lines to ~300 lines through proper separation of concerns.
 */
public class TaskActivity extends Activity implements
        TaskListAdapter.TaskActionListener,
        TaskEditDialog.TaskSaveListener,
        CompletionDialog.CompletionListener {

    private static final String TAG = "TaskActivity";

    // Data components
    private TaskDao taskDao;
    private CompletionDao completionDao;
    private StatsManager statsManager;
    private AppLogger logger;

    // UI components
    private ListView taskListView;
    private TextView emptyText;
    private TextView statsText;
    private EditText searchInput;
    private Spinner categorySpinner;
    private Spinner completionSpinner;
    private Spinner sortSpinner;
    private Button addTaskButton;

    // Adapters and dialogs
    private TaskListAdapter taskAdapter;
    private TaskEditDialog editDialog;
    private CompletionDialog completionDialog;

    // Data
    private List<Task> allTaskList = new ArrayList<>();
    private List<Task> filteredTaskList = new ArrayList<>();
    private List<String> categories = new ArrayList<>();

    // Filter and sort state
    private int selectedCategory = 0;
    private int completionFilter = 0;
    private int sortOption = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Initialize components
        initializeComponents();

        // Setup UI
        setupViews();
        setupFilters();
        setupSearch();

        // Load initial data
        loadTasks();
    }

    /**
     * Initialize data components
     */
    private void initializeComponents() {
        logger = AppLogger.getInstance(this);
        taskDao = new TaskDao(this);
        completionDao = new CompletionDao(this);
        statsManager = new StatsManager(this);

        editDialog = new TaskEditDialog(this, this);
        completionDialog = new CompletionDialog(this, this);

        logger.info(TAG, "TaskActivity initialized with refactored architecture");
    }

    /**
     * Setup UI views
     */
    private void setupViews() {
        taskListView = findViewById(R.id.taskListView);
        emptyText = findViewById(R.id.emptyTaskText);
        statsText = findViewById(R.id.statsText);
        searchInput = findViewById(R.id.searchInput);
        categorySpinner = findViewById(R.id.categoryFilter);
        completionSpinner = findViewById(R.id.completionFilter);
        sortSpinner = findViewById(R.id.sortSpinner);
        addTaskButton = findViewById(R.id.addTaskButton);

        // Setup task list adapter
        taskAdapter = new TaskListAdapter(this, filteredTaskList, this);
        taskListView.setAdapter(taskAdapter);

        // Add task button
        addTaskButton.setOnClickListener(v -> editDialog.showAddDialog());
    }

    /**
     * Setup filter spinners
     */
    private void setupFilters() {
        // Category filter
        categorySpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                selectedCategory = position;
                applyFilters();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });

        // Completion filter
        String[] completionOptions = {"All Tasks", "Active Only", "Completed Only"};
        completionSpinner.setAdapter(new ArrayAdapter<>(this,
                android.R.layout.simple_spinner_dropdown_item, completionOptions));

        completionSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                completionFilter = position;
                applyFilters();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });

        // Sort options
        String[] sortOptions = {
            "Priority (High to Low)",
            "Due Date (Nearest First)",
            "Created (Newest First)",
            "Title (A-Z)",
            "Category"
        };
        sortSpinner.setAdapter(new ArrayAdapter<>(this,
                android.R.layout.simple_spinner_dropdown_item, sortOptions));

        sortSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                sortOption = position;
                applySorting();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }

    /**
     * Setup search functionality
     */
    private void setupSearch() {
        searchInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                applyFilters();
            }
        });
    }

    /**
     * Load tasks from database
     */
    private void loadTasks() {
        allTaskList = taskDao.getAllTasks();
        updateCategoryFilter();
        applyFilters();
        updateStatistics();

        logger.debug(TAG, "Loaded " + allTaskList.size() + " tasks");
    }

    /**
     * Update category filter options
     */
    private void updateCategoryFilter() {
        categories.clear();
        categories.add("All Categories");
        categories.addAll(taskDao.getAllCategories());

        ArrayAdapter<String> adapter = new ArrayAdapter<>(this,
                android.R.layout.simple_spinner_dropdown_item, categories);
        categorySpinner.setAdapter(adapter);
    }

    /**
     * Apply filters to task list
     */
    private void applyFilters() {
        filteredTaskList.clear();
        String searchQuery = searchInput.getText().toString().toLowerCase().trim();

        for (Task task : allTaskList) {
            // Apply completion filter
            if (completionFilter == 1 && task.isCompleted()) continue;
            if (completionFilter == 2 && !task.isCompleted()) continue;

            // Apply category filter
            if (selectedCategory > 0) {
                String selectedCat = categories.get(selectedCategory);
                if (task.getCategory() == null || !task.getCategory().equals(selectedCat)) {
                    continue;
                }
            }

            // Apply search filter
            if (!searchQuery.isEmpty()) {
                boolean matches = task.getTitle().toLowerCase().contains(searchQuery) ||
                        (task.getDescription() != null &&
                         task.getDescription().toLowerCase().contains(searchQuery));
                if (!matches) continue;
            }

            filteredTaskList.add(task);
        }

        applySorting();
    }

    /**
     * Apply sorting to filtered list
     */
    private void applySorting() {
        Collections.sort(filteredTaskList, (task1, task2) -> {
            switch (sortOption) {
                case 0: // Priority (High to Low)
                    return Integer.compare(task2.getPriority(), task1.getPriority());
                case 1: // Due Date (Nearest First)
                    if (task1.getDueDate() == 0) return 1;
                    if (task2.getDueDate() == 0) return -1;
                    return Long.compare(task1.getDueDate(), task2.getDueDate());
                case 2: // Created (Newest First)
                    return Long.compare(task2.getCreatedAt(), task1.getCreatedAt());
                case 3: // Title (A-Z)
                    return task1.getTitle().compareToIgnoreCase(task2.getTitle());
                case 4: // Category
                    String cat1 = task1.getCategory() != null ? task1.getCategory() : "";
                    String cat2 = task2.getCategory() != null ? task2.getCategory() : "";
                    return cat1.compareToIgnoreCase(cat2);
                default:
                    return 0;
            }
        });

        updateUI();
    }

    /**
     * Update UI after data changes
     */
    private void updateUI() {
        taskAdapter.notifyDataSetChanged();
        emptyText.setVisibility(filteredTaskList.isEmpty() ? View.VISIBLE : View.GONE);
        taskListView.setVisibility(filteredTaskList.isEmpty() ? View.GONE : View.VISIBLE);
    }

    /**
     * Update statistics display
     */
    private void updateStatistics() {
        statsManager.calculateStatistics(allTaskList);
        statsText.setText(statsManager.getStatsSummary());
    }

    // TaskListAdapter.TaskActionListener implementation

    @Override
    public void onTaskCheckedChanged(Task task, boolean isChecked) {
        if (isChecked) {
            completionDialog.show(task);
        } else {
            task.setCompleted(false);
            taskDao.markTaskCompleted(task.getId(), false);
            logger.info(TAG, "Task " + task.getTitle() + " marked as active");
            loadTasks();
        }
    }

    @Override
    public void onTaskEdit(Task task) {
        editDialog.showEditDialog(task);
    }

    @Override
    public void onTaskDelete(Task task) {
        taskDao.deleteTask(task.getId());
        completionDao.deleteCompletionsForTask(task.getId());
        logger.info(TAG, "Task deleted: " + task.getTitle());
        Toast.makeText(this, "Task deleted", Toast.LENGTH_SHORT).show();
        loadTasks();
    }

    // TaskEditDialog.TaskSaveListener implementation

    @Override
    public void onTaskSaved(Task task) {
        String message = task.getId() == 0 ? "Task created" : "Task updated";
        if (task.isRecurring()) {
            message += " - " + task.getRecurrenceString();
        }
        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
        loadTasks();
    }

    // CompletionDialog.CompletionListener implementation

    @Override
    public void onTaskCompleted(Task task) {
        Toast.makeText(this, "Task completed!", Toast.LENGTH_SHORT).show();
        loadTasks();
    }

    @Override
    public void onCompletionCancelled(Task task) {
        // Reset checkbox state
        loadTasks();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        taskDao.close();
        completionDao.close();
        statsManager.close();
    }
}

File: app/src/main/java/com/secretary/features/tasks/presentation/activity/TaskActivity.kt
package com.secretary.features.tasks.presentation.activity

import com.secretary.R
import com.secretary.features.tasks.domain.model.Task
import com.secretary.features.tasks.presentation.util.TaskFilterManager
import com.secretary.features.tasks.presentation.adapter.TaskListAdapter
import com.secretary.core.logging.AppLogger
import com.secretary.features.motivation.presentation.StreakColorUtil
import com.secretary.features.statistics.data.CompletionRepositoryImpl
import com.secretary.features.statistics.domain.model.TaskStatistics
import com.secretary.features.statistics.domain.repository.CompletionRepository
import com.secretary.features.tasks.data.TaskDao
import com.secretary.features.tasks.domain.repository.TaskRepository
import com.secretary.features.tasks.presentation.viewmodel.TaskListViewModel
import com.secretary.features.tasks.presentation.viewmodel.TaskViewModelFactory
import com.secretary.shared.database.TaskDatabase
import androidx.appcompat.app.AppCompatActivity
import android.app.AlertDialog
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.*
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import com.google.android.material.progressindicator.CircularProgressIndicator
import kotlinx.coroutines.launch
import java.util.*

/**
 * Task Activity - Main Task Management UI
 * Phase 4.5.6: Dialog Extraction - Uses DialogFragments with MVVM
 * Phase 4.5.3 Wave 9: Converted to Kotlin
 *
 * Displays task list with search, filtering, and sorting.
 * Uses AddTaskDialog, EditTaskDialog, CompletionDialog (DialogFragments).
 * Implements TaskActionListener for adapter callbacks.
 */
class TaskActivity : AppCompatActivity(), TaskListAdapter.TaskActionListener {

    companion object {
        private const val TAG = "TaskActivity"
    }

    // Dependencies
    private lateinit var repository: TaskRepository // NEW - for Task CRUD operations
    private lateinit var viewModel: TaskListViewModel // ViewModel for MVVM pattern
    private lateinit var filterManager: TaskFilterManager
    private lateinit var viewModelFactory: TaskViewModelFactory // Factory for DialogFragments

    // Views
    private lateinit var taskListView: ListView
    private lateinit var emptyTasksText: TextView
    private lateinit var addTaskButton: Button
    private lateinit var searchEditText: EditText
    private lateinit var statusFilterSpinner: Spinner
    private lateinit var priorityFilterSpinner: Spinner
    private lateinit var categoryFilterSpinner: Spinner
    private lateinit var sortBySpinner: Spinner

    // Motivation UI components
    private lateinit var flameIcon: ImageView
    private lateinit var streakCurrentText: TextView
    private lateinit var streakBestText: TextView
    private lateinit var streakProgressIndicator: CircularProgressIndicator
    private lateinit var todayProgressIndicator: CircularProgressIndicator
    private lateinit var todayPercentText: TextView
    private lateinit var todayDetailText: TextView
    private lateinit var weekProgressIndicator: CircularProgressIndicator
    private lateinit var weekPercentText: TextView
    private lateinit var weekDetailText: TextView
    private lateinit var motivationalMessageText: TextView

    // Data
    private lateinit var adapter: TaskListAdapter
    private val taskList = ArrayList<Task>()
    private val filteredTaskList = ArrayList<Task>()
    private var allCategories = listOf<String>()

    // Filter states
    private var searchQuery = ""
    private var statusFilter = 0        // 0=All, 1=Active, 2=Completed
    private var priorityFilter = -1     // -1=All, 0=Low, 1=Medium, 2=High
    private var categoryFilter: String? = null
    private var sortOption = 0          // Index in TaskFilterManager.SortOption enum

    // ========== Lifecycle Methods ==========

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        try {
            setContentView(R.layout.activity_tasks)
            AppLogger.initialize(this)
            AppLogger.info(TAG, "TaskActivity started - setContentView successful")

            // Initialize Room database and repositories
            val database = TaskDatabase.getDatabase(this)
            val taskDao = database.taskDao()
            repository = com.secretary.features.tasks.data.repository.TaskRepositoryImpl(taskDao)

            // Phase 4: Initialize CompletionRepository for statistics
            val completionDao = database.completionDao()
            val completionRepository: CompletionRepository = CompletionRepositoryImpl(completionDao)

            // Initialize Services for domain logic
            val streakService = com.secretary.features.tasks.domain.service.StreakService()
            val recurrenceService = com.secretary.features.tasks.domain.service.RecurrenceService()

            // Initialize MotivationalMessageService for DI
            val motivationalMessageService = com.secretary.features.motivation.domain.MotivationalMessageService()

            // Initialize ViewModel with Factory (dependency injection)
            viewModelFactory = com.secretary.features.tasks.presentation.viewmodel.TaskViewModelFactory(
                repository,
                completionRepository,
                streakService,
                recurrenceService,
                motivationalMessageService
            )
            viewModel = androidx.lifecycle.ViewModelProvider(this, viewModelFactory)
                .get(com.secretary.features.tasks.presentation.viewmodel.TaskListViewModel::class.java)

            // Setup FragmentResult listeners for DialogFragments
            setupFragmentResultListeners()

            // Setup ViewModel observers (MVVM pattern)
            setupViewModelObservers()

            // Find views
            taskListView = findViewById(R.id.taskListView)
            emptyTasksText = findViewById(R.id.emptyTasksText)
            addTaskButton = findViewById(R.id.addTaskButton)
            searchEditText = findViewById(R.id.searchEditText)
            statusFilterSpinner = findViewById(R.id.statusFilterSpinner)
            priorityFilterSpinner = findViewById(R.id.priorityFilterSpinner)
            categoryFilterSpinner = findViewById(R.id.categoryFilterSpinner)
            sortBySpinner = findViewById(R.id.sortBySpinner)

            // Find motivation UI components
            flameIcon = findViewById(R.id.flameIcon)
            streakCurrentText = findViewById(R.id.streakCurrentText)
            streakBestText = findViewById(R.id.streakBestText)
            streakProgressIndicator = findViewById(R.id.streakProgressIndicator)
            todayProgressIndicator = findViewById(R.id.todayProgressIndicator)
            todayPercentText = findViewById(R.id.todayPercentText)
            todayDetailText = findViewById(R.id.todayDetailText)
            weekProgressIndicator = findViewById(R.id.weekProgressIndicator)
            weekPercentText = findViewById(R.id.weekPercentText)
            weekDetailText = findViewById(R.id.weekDetailText)
            motivationalMessageText = findViewById(R.id.motivationalMessageText)

            AppLogger.info(TAG, "All views found successfully")

            // Setup adapter
            adapter = TaskListAdapter(this, filteredTaskList, this)
            taskListView.adapter = adapter

            // Setup filter manager
            filterManager = TaskFilterManager()

            // Setup UI
            setupFilterSpinners()
            setupSearch()

            // Add task button
            addTaskButton.setOnClickListener {
                AppLogger.info(TAG, "Add task button clicked")
                showAddTaskDialog()
            }

            // Initial load
            loadTasks()

            AppLogger.info(TAG, "TaskActivity onCreate completed successfully")

        } catch (e: Exception) {
            android.util.Log.e(TAG, "FATAL ERROR in TaskActivity.onCreate()", e)
            e.printStackTrace() // Print full stack trace to logcat

            try {
                AppLogger.initialize(this) // Ensure logger is initialized
                AppLogger.error(TAG, "TaskActivity onCreate crashed: ${e.javaClass.name}: ${e.message}", e)
            } catch (logErr: Exception) {
                android.util.Log.e(TAG, "Could not log error", logErr)
            }

            // Show error dialog with full stack trace
            val stackTrace = e.stackTraceToString()
            AlertDialog.Builder(this)
                .setTitle("TaskActivity Crash")
                .setMessage("Error: ${e.javaClass.simpleName}\n${e.message}\n\nStack trace:\n${stackTrace.take(500)}")
                .setPositiveButton("Close App") { _, _ ->
                    finish()
                    android.os.Process.killProcess(android.os.Process.myPid())
                }
                .setCancelable(false)
                .show()
        }
    }

    /**
     * Setup listeners for TaskDialogHelper callbacks
     */
    /**
     * Setup FragmentResult listeners for DialogFragments
     * Phase 4.5.6: Replaces setupDialogHelperListeners()
     */
    private fun setupFragmentResultListeners() {
        // Listen for AddTaskDialog results
        supportFragmentManager.setFragmentResultListener(
            com.secretary.features.tasks.presentation.dialog.AddTaskDialog.RESULT_KEY,
            this
        ) { _, _ ->
            // Task was saved - reload task list
            loadTasks()
        }

        // Listen for EditTaskDialog results
        supportFragmentManager.setFragmentResultListener(
            com.secretary.features.tasks.presentation.dialog.EditTaskDialog.RESULT_KEY,
            this
        ) { _, _ ->
            // Task was updated - reload task list
            loadTasks()
        }

        // Listen for CompletionDialog results
        supportFragmentManager.setFragmentResultListener(
            com.secretary.features.tasks.presentation.dialog.CompletionDialog.RESULT_KEY,
            this
        ) { _, bundle ->
            // Task was completed or cancelled - reload task list
            val cancelled = bundle.getBoolean("cancelled", false)
            if (cancelled) {
                // User cancelled - reload to reset UI
                loadTasks()
            } else {
                // Task completed - reload task list
                loadTasks()
            }
        }
    }

    /**
     * Setup LiveData observers for ViewModel (MVVM pattern)
     */
    private fun setupViewModelObservers() {
        // Observe tasks - update task list and apply filters
        viewModel.tasks.observe(this) { tasks ->
            taskList.clear()
            taskList.addAll(tasks)
            lifecycleScope.launch {
                updateCategoryFilter() // Update category filter with new categories
            }
            applyFilters() // Apply current filters to show filtered list
            AppLogger.info(TAG, "ViewModel: Loaded ${tasks.size} tasks")
        }

        // Phase 4: Observe statistics - update statistics display
        viewModel.statistics.observe(this) { stats ->
            stats?.let {
                AppLogger.info(TAG, "Statistics updated: ${it.toDisplayString()}")
                updateStreakIndicator(it.highestActiveStreak, it.longestStreakEver)
                updateCompletionRates(it)
            }
        }

        // Observe motivational message - update motivational message display
        viewModel.motivationalMessage.observe(this) { message ->
            motivationalMessageText.text = message
        }

        // Observe error - show error Toast
        viewModel.error.observe(this) { errorMessage ->
            errorMessage?.let {
                Toast.makeText(this, "Error: $it", Toast.LENGTH_LONG).show()
                AppLogger.error(TAG, "ViewModel Error: $it")
                viewModel.clearError() // Clear after showing
            }
        }

        // Observe operation success - show success Toast
        viewModel.operationSuccess.observe(this) { successMessage ->
            successMessage?.let {
                Toast.makeText(this, it, Toast.LENGTH_SHORT).show()
                AppLogger.info(TAG, "ViewModel Success: $it")
                viewModel.clearOperationSuccess() // Clear after showing
            }
        }
    }

    /**
     * Setup filter spinners with adapters and listeners
     */
    private fun setupFilterSpinners() {
        // Status filter
        val statusOptions = arrayOf("All Tasks", "Active Only", "Completed Only")
        val statusAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, statusOptions).apply {
            setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        }
        statusFilterSpinner.adapter = statusAdapter
        statusFilterSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                statusFilter = position
                applyFilters()
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }

        // Priority filter
        val priorityOptions = arrayOf("All Priorities", "Low", "Medium", "High")
        val priorityAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, priorityOptions).apply {
            setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        }
        priorityFilterSpinner.adapter = priorityAdapter
        priorityFilterSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                priorityFilter = position - 1 // -1 = All, 0 = Low, 1 = Medium, 2 = High
                applyFilters()
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }

        // Sort by
        val sortOptions = arrayOf("Sort: Priority", "Sort: Due Date", "Sort: Category", "Sort: Created", "Sort: Title")
        val sortAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, sortOptions).apply {
            setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        }
        sortBySpinner.adapter = sortAdapter
        sortBySpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                sortOption = position
                applyFilters() // Use applyFilters which includes sorting
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    /**
     * Setup search text watcher
     */
    private fun setupSearch() {
        searchEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                searchQuery = s.toString().lowercase(Locale.getDefault())
                applyFilters()
            }

            override fun afterTextChanged(s: Editable?) {}
        })
    }

    /**
     * Apply current filters and sorting to task list
     */
    private fun applyFilters() {
        // Configure filter manager based on current UI state
        filterManager.searchQuery = searchQuery
        filterManager.categoryFilter = categoryFilter

        // Convert status filter to enum
        val completionFilter = when (statusFilter) {
            1 -> TaskFilterManager.CompletionFilter.ACTIVE_ONLY
            2 -> TaskFilterManager.CompletionFilter.COMPLETED_ONLY
            else -> TaskFilterManager.CompletionFilter.ALL
        }
        filterManager.completionFilter = completionFilter

        // Convert sort option to enum
        val sortOptions = TaskFilterManager.SortOption.values()
        if (sortOption in sortOptions.indices) {
            filterManager.sortOption = sortOptions[sortOption]
        }

        // Apply filters and sorting
        filteredTaskList.clear()
        filteredTaskList.addAll(filterManager.applyFilters(taskList))
        filterManager.sortTasks(filteredTaskList)

        // Show/hide empty view
        if (filteredTaskList.isEmpty()) {
            taskListView.visibility = View.GONE
            emptyTasksText.visibility = View.VISIBLE
            if (searchQuery.isNotEmpty() || statusFilter > 0 || priorityFilter >= 0) {
                emptyTasksText.text = "No tasks match your filters."
            } else {
                emptyTasksText.text = "No tasks yet.\nTap + to add a task."
            }
        } else {
            taskListView.visibility = View.VISIBLE
            emptyTasksText.visibility = View.GONE
        }

        adapter.notifyDataSetChanged()
        AppLogger.info(TAG, "Filters applied: ${filteredTaskList.size} tasks shown")
    }

    /**
     * Load all tasks from database via ViewModel
     * The ViewModel observer will handle updating the UI automatically
     */
    private fun loadTasks() {
        viewModel.loadTasks()
    }

    /**
     * Update statistics display at top of screen using Repository
     * TODO: Update to use new motivation UI components
     */
    private suspend fun updateStatistics() {
        try {
            val todayCount = repository.getTasksCompletedToday()
            val weekCount = repository.getTasksCompletedLast7Days()
            val overdueCount = repository.getOverdueTasksCount()

            // TODO: Update streak indicator, completion rates, and motivational message
            AppLogger.info(TAG, "Statistics: Today=$todayCount, Week=$weekCount, Overdue=$overdueCount")
        } catch (e: Exception) {
            AppLogger.error(TAG, "Failed to update statistics from Repository", e)
        }
    }

    /**
     * Update category filter spinner with all categories from database using Repository
     */
    private suspend fun updateCategoryFilter() {
        try {
            // Get all unique categories from database
            allCategories = repository.getAllCategories()

            // Build options for category filter spinner
            val categoryOptions = Array(allCategories.size + 1) { i ->
                if (i == 0) "All Categories" else allCategories[i - 1]
            }

            val categoryAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, categoryOptions).apply {
                setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            }
            categoryFilterSpinner.adapter = categoryAdapter

            // Restore previous selection if possible
            categoryFilter?.let { filter ->
                val index = allCategories.indexOf(filter) + 1
                if (index > 0 && index < categoryOptions.size) {
                    categoryFilterSpinner.setSelection(index)
                }
            }

            categoryFilterSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                    categoryFilter = if (position == 0) null else allCategories[position - 1]
                    applyFilters()
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
        } catch (e: Exception) {
            AppLogger.error(TAG, "Failed to update category filter from Repository", e)
        }
    }

    /**
     * Show dialog for adding a new task
     * Phase 4.5.6: Uses AddTaskDialog DialogFragment
     */
    private fun showAddTaskDialog() {
        val dialog = com.secretary.features.tasks.presentation.dialog.AddTaskDialog.newInstance(
            allCategories,
            viewModelFactory
        )
        dialog.show(supportFragmentManager, com.secretary.features.tasks.presentation.dialog.AddTaskDialog.TAG)
    }

    /**
     * Show dialog for editing an existing task
     * Phase 4.5.6: Uses EditTaskDialog DialogFragment
     */
    private fun showEditTaskDialog(existingTask: Task) {
        val dialog = com.secretary.features.tasks.presentation.dialog.EditTaskDialog.newInstance(
            existingTask.id,
            allCategories,
            viewModelFactory
        )
        dialog.show(supportFragmentManager, com.secretary.features.tasks.presentation.dialog.EditTaskDialog.TAG)
    }

    /**
     * Show completion dialog with time tracking
     * Phase 4.5.6: Uses CompletionDialog DialogFragment
     */
    private fun showCompletionDialog(task: Task) {
        val dialog = com.secretary.features.tasks.presentation.dialog.CompletionDialog.newInstance(
            task,
            viewModelFactory
        )
        dialog.show(supportFragmentManager, com.secretary.features.tasks.presentation.dialog.CompletionDialog.TAG)
    }

    override fun onResume() {
        super.onResume()
        loadTasks()
    }

    // Note: Room database is managed automatically by TaskDatabase.getDatabase() singleton
    // No need to manually close database connections

    // ========== TaskActionListener Interface Implementation ==========

    override fun onTaskCheckChanged(task: Task, isChecked: Boolean) {
        AppLogger.info(TAG, "Task checkbox changed: ${task.title} -> $isChecked")

        if (isChecked) {
            // Show completion dialog
            showCompletionDialog(task)
        } else {
            // Unchecked - mark as incomplete
            task.isCompleted = false
            viewModel.updateTask(task)
        }
    }

    override fun onTaskEdit(task: Task) {
        AppLogger.info(TAG, "Edit task: ${task.title}")
        showEditTaskDialog(task)
    }

    override fun onTaskDelete(task: Task) {
        AppLogger.info(TAG, "Delete task: ${task.title}")
        viewModel.deleteTask(task.id)
    }

    override fun onTasksChanged() {
        AppLogger.info(TAG, "Tasks changed - reloading")
        loadTasks()
    }

    // ========== Helper Methods for Motivation UI ==========

    /**
     * Update the streak indicator with current and best streak values
     * Updates the flame icon color, progress indicator, and text displays
     *
     * @param current Current active streak in days
     * @param best Best (longest) streak ever achieved
     */
    private fun updateStreakIndicator(current: Int, best: Int) {
        streakCurrentText.text = "Current: $current days"
        streakBestText.text = "Best: $best days"

        val progress = if (best > 0) (current * 100 / best) else 0
        streakProgressIndicator.progress = progress

        val colorRes = StreakColorUtil.getStreakColor(current)
        flameIcon.setColorFilter(ContextCompat.getColor(this, colorRes))
        streakProgressIndicator.setIndicatorColor(ContextCompat.getColor(this, colorRes))
    }

    /**
     * Update completion rate displays for today and this week
     * Updates progress indicators, percentages, and detail text
     *
     * @param stats TaskStatistics containing completion data
     */
    private fun updateCompletionRates(stats: TaskStatistics) {
        // Today
        val todayPercent = stats.getTodayCompletionRate()
        todayProgressIndicator.progress = todayPercent
        todayPercentText.text = "$todayPercent%"
        todayDetailText.text = "${stats.completedToday} of ${stats.totalTasks} tasks"
        val todayColor = StreakColorUtil.getCompletionColor(todayPercent)
        todayProgressIndicator.setIndicatorColor(ContextCompat.getColor(this, todayColor))

        // Week
        val weekPercent = stats.getWeekCompletionRate()
        weekProgressIndicator.progress = weekPercent
        weekPercentText.text = "$weekPercent%"
        weekDetailText.text = "${stats.completedThisWeek} of ${stats.totalTasks} tasks"
        val weekColor = StreakColorUtil.getCompletionColor(weekPercent)
        weekProgressIndicator.setIndicatorColor(ContextCompat.getColor(this, weekColor))
    }
}
